File: 2800-2899/2812. Find the Safest Path in a Grid/Solution.java
import java.util.*;

class Solution {
    int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};

    public int maximumSafenessFactor(List<List<Integer>> grid) {
        int n = grid.size();
        int[][] dist = new int[n][n];
        for (int[] row : dist) Arrays.fill(row, -1);

        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (grid.get(i).get(j) == 1) {
                    dist[i][j] = 0;
                    queue.offer(new int[]{i, j});
                }
            }
        }

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            for (int[] d : dirs) {
                int nr = cur[0] + d[0], nc = cur[1] + d[1];
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && dist[nr][nc] == -1) {
                    dist[nr][nc] = dist[cur[0]][cur[1]] + 1;
                    queue.offer(new int[]{nr, nc});
                }
            }
        }

        if (dist[0][0] == 0 || dist[n-1][n-1] == 0) return 0;

        int lo = 0, hi = 2 * n, result = 0;
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            if (canReach(dist, n, mid)) {
                result = mid;
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        return result;
    }

    private boolean canReach(int[][] dist, int n, int k) {
        if (dist[0][0] < k || dist[n-1][n-1] < k) return false;

        boolean[][] visited = new boolean[n][n];
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0});
        visited[0][0] = true;

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int r = cur[0], c = cur[1];
            if (r == n-1 && c == n-1) return true;
            for (int[] d : dirs) {
                int nr = r + d[0], nc = c + d[1];
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && !visited[nr][nc] && dist[nr][nc] >= k) {
                    visited[nr][nc] = true;
                    queue.offer(new int[]{nr, nc});
                }
            }
        }

        return false;
    }
}

File: 2800-2899/2812. Find the Safest Path in a Grid/Solution.cpp
//Approach-1 (Multi-Source BFS + Binary Search + BFS)
//T.C : O(n² log n)
//S.C : O(n²)

class Solution {
public:

    vector<vector<int>> directions{
        {0, 1},
        {0, -1},
        {1, 0},
        {-1, 0}
    };

    bool canReach(
        vector<vector<int>>& distance,
        int n,
        int limit
    ) {

        if (
            distance[0][0] < limit ||
            distance[n - 1][n - 1] < limit
        ) {
            return false;
        }

        vector<vector<bool>> visited(
            n,
            vector<bool>(n, false)
        );

        queue<pair<int, int>> bfsQueue;

        bfsQueue.push({0, 0});

        visited[0][0] = true;

        while (!bfsQueue.empty()) {

            auto current = bfsQueue.front();

            bfsQueue.pop();

            int row = current.first;

            int col = current.second;

            if (
                row == n - 1 &&
                col == n - 1
            ) {
                return true;
            }

            for (auto& direction : directions) {

                int nextRow = row + direction[0];

                int nextCol = col + direction[1];

                if (
                    nextRow >= 0 &&
                    nextRow < n &&
                    nextCol >= 0 &&
                    nextCol < n &&
                    !visited[nextRow][nextCol] &&
                    distance[nextRow][nextCol] >= limit
                ) {

                    visited[nextRow][nextCol] = true;

                    bfsQueue.push(
                        {
                            nextRow,
                            nextCol
                        }
                    );
                }
            }
        }

        return false;
    }

    int maximumSafenessFactor(
        vector<vector<int>>& grid
    ) {

        int n = grid.size();

        vector<vector<int>> distance(
            n,
            vector<int>(n, -1)
        );

        queue<pair<int, int>> bfsQueue;

        for (int row = 0; row < n; row++) {

            for (int col = 0; col < n; col++) {

                if (grid[row][col] == 1) {

                    distance[row][col] = 0;

                    bfsQueue.push(
                        {
                            row,
                            col
                        }
                    );
                }
            }
        }

        while (!bfsQueue.empty()) {

            auto current = bfsQueue.front();

            bfsQueue.pop();

            for (auto& direction : directions) {

                int nextRow =
                    current.first + direction[0];

                int nextCol =
                    current.second + direction[1];

                if (
                    nextRow >= 0 &&
                    nextRow < n &&
                    nextCol >= 0 &&
                    nextCol < n &&
                    distance[nextRow][nextCol] == -1
                ) {

                    distance[nextRow][nextCol] =
                        distance[current.first][current.second] + 1;

                    bfsQueue.push(
                        {
                            nextRow,
                            nextCol
                        }
                    );
                }
            }
        }

        if (
            distance[0][0] == 0 ||
            distance[n - 1][n - 1] == 0
        ) {
            return 0;
        }

        int low = 0;

        int high = 2 * n;

        int answer = 0;

        while (low <= high) {

            int middle = (low + high) / 2;

            if (
                canReach(
                    distance,
                    n,
                    middle
                )
            ) {

                answer = middle;

                low = middle + 1;

            } else {

                high = middle - 1;
            }
        }

        return answer;
    }
};